fpdf/drawing.py

```diff
@@ -0,0 +1,4207 @@
+    is similar to shining a harsh spotlight on the backdrop."
+    """
+    SOFT_LIGHT = Name("SoftLight")
+    """
+    "Darkens or lightens the colors, depending on the source color value. The effect is
+    similar to shining a diffused spotlight on the backdrop."
+    """
+    DIFFERENCE = Name("Difference")
+    '''"Subtracts the darker of the two constituent colors from the lighter color."'''
+    EXCLUSION = Name("Exclusion")
+    """
+    "Produces an effect similar to that of the Difference mode but lower in contrast.
+    Painting with white inverts the backdrop color; painting with black produces no
+    change."
+    """
+
+    HUE = Name("Hue")
+    """
+    "Creates a color with the hue of the source color and the saturation and luminosity
+    of the backdrop color."
+    """
+    SATURATION = Name("Saturation")
+    """
+    "Creates a color with the saturation of the source color and the hue and luminosity
+    of the backdrop color. Painting with this mode in an area of the backdrop that is
+    a pure gray (no saturation) produces no change."
+    """
+    COLOR = Name("Color")
+    """
+    "Creates a color with the hue and saturation of the source color and the luminosity
+    of the backdrop color. This preserves the gray levels of the backdrop and is
+    useful for coloring monochrome images or tinting color images."
+    """
+    LUMINOSITY = Name("Luminosity")
+    """
+    "Creates a color with the luminosity of the source color and the hue and saturation
+    of the backdrop color. This produces an inverse effect to that of the Color mode."
+    """
+
+
+class PDFStyleKeys(enum.Enum):
+    """
+    An enumeration of the graphics state parameter dictionary keys.
+    """
+
+    FILL_ALPHA = Name("ca")
+    BLEND_MODE = Name("BM")  # shared between stroke and fill
+    STROKE_ALPHA = Name("CA")
+    STROKE_ADJUSTMENT = Name("SA")
+    STROKE_WIDTH = Name("LW")
+    STROKE_CAP_STYLE = Name("LC")
+    STROKE_JOIN_STYLE = Name("LJ")
+    STROKE_MITER_LIMIT = Name("ML")
+    STROKE_DASH_PATTERN = Name("D")  # array of array, number, e.g. [[1 1] 0]
+
+
+class GraphicsStyle:
+    """
+    A class representing various style attributes that determine drawing appearance.
+
+    This class uses the convention that the global Python singleton ellipsis (`...`) is
+    exclusively used to represent values that are inherited from the parent style. This
+    is to disambiguate the value None which is used for several values to signal an
+    explicitly disabled style. An example of this is the fill/stroke color styles,
+    which use None as hints to the auto paint style detection code.
+    """
+
+    INHERIT = ...
+    """Singleton specifying a style parameter should be inherited from the parent context."""
+
+    # order is be important here because some of these properties are entangled, e.g.
+    # stroke_dash_pattern and stroke_dash_phase
+    MERGE_PROPERTIES = (
```

Comment:
> Do you think those 2 classes should be merged or related, for exemple one including an instance of the other?

summary: I do think the two implementations could be merged, but I don't think it would be trivial to do so, and I'd rather not try to fit that work into this PR.

The two take different approaches to storing their stacks. The `GraphicsStateMixin` stores all graphics graphics stack levels internally, while the `GraphicsStyle` only stores a single level, with each of `PaintedPath`, `ClippingPath` and `GraphicsContext` storing one level of the style, with the resolved style not emitted until the entire drawing is being serialized/output. This is more of a compositional design because I've struggled with Python's multiple inheritance in the past and prefer to avoid it when possible.

> Also, when creating a `new_path`, does the default `GraphicsStyle` takes values of `fill_color` / `draw_color` / `line_width` from the `FPDF` instance?

The answer to this is "yes, but". This should work as expected in almost all cases (I'll add some unit tests to verify this behavior as well), however there is an awkward case around the automatic paint style detection. Currently, inheriting the `fill_color` etc. from the parent `FPDF` is done implicitly, i.e. the structure of the output PDF causes the drawing/path to inherit the applicable values of the document's graphics state, without any extra work on the Python side. However, the automatic paint style detection (which decides stroke-only, fill-only or stroke+fill) depends on explicit information being captured within the Python data structures. The problem mainly is that the `FPDF` object sets defaults for the stroke/fill values even if they aren't set explicitly by the user, and there is no way to unset them, so the auto detection would always resolve to stroke+fill if we explicitly passed the document values to the base `DrawingContext` object.

The auto detection mostly exists to simplify the SVG converter, but it is also a definitely a convenience for hand-written paths as well. The core issue is that a 0-width stroke in a PDF is actually a hairline stroke rather than no stroke at all. Without the autodetection, there are two possibly conflicting pieces of information that define how the path looks: both setting the `stroke_width` and the `paint_rule`. The autodetection makes for a simpler API where the user can define how the path looks by only setting the `stroke_width` and not having to worry about also remembering to set the `paint_rule`.

Reply:
I've addressed the double scaling in a two-parter: b61ce6ed040363aab892a6d93fa168e01c8083ab and 6011b92480c89407888d1db985e860a17ddc9f67. This is a little bit of a kludge, but the "correct" solution, which in my opinion would be to set a user-unit-to-pdf-unit transform at the beginning of every page, would be a much more invasive change that would require regenerating all of the test PDFs. The solution I chose only required localized changes around the drawing API and the way that the `dash_pattern` is stored in the `GraphicsStateMixin`. The default behavior of `PathPaintRule.AUTO` actually appears to work better than I thought it did, so I don't think I need to change anything related to that.